osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/VerifyReplayExistsCommand.cs

```diff
@@ -0,0 +1,108 @@
+        /// </summary>
+        [Option(CommandOptionType.SingleValue, Template = "--start-id")]
+        public ulong? StartId { get; set; }
+
+        [Option(CommandOptionType.SingleOrNoValue, Template = "-v|--verbose", Description = "Output when a score is preserved too.")]
+        public bool Verbose { get; set; }
+
+        /// <summary>
+        /// The number of scores to run in each batch. Setting this higher will cause larger SQL statements for insert.
+        /// </summary>
+        [Option(CommandOptionType.SingleValue, Template = "--batch-size")]
+        public int BatchSize { get; set; } = 5000;
+
+        [Option(CommandOptionType.SingleOrNoValue, Template = "--dry-run")]
+        public bool DryRun { get; set; }
+
+        public async Task<int> OnExecuteAsync(CancellationToken cancellationToken)
+        {
+            ulong lastId = StartId ?? 0;
+
+            var s3Key = Environment.GetEnvironmentVariable("S3_KEY") ?? string.Empty;
+            var s3Secret = Environment.GetEnvironmentVariable("S3_SECRET") ?? string.Empty;
+
+            using var s3 = new AmazonS3Client(new BasicAWSCredentials(s3Key, s3Secret), new AmazonS3Config
+            {
+                CacheHttpClient = true,
+                HttpClientCacheSize = 32,
+                RegionEndpoint = RegionEndpoint.USWest1,
+                UseHttp = true,
+                ForcePathStyle = true
+            });
+
+            using var conn = DatabaseAccess.GetConnection();
+
+            Console.WriteLine();
+            Console.WriteLine($"Verifying replay flag for scores starting from {lastId}");
+
+            if (DryRun)
+                Console.WriteLine("RUNNING IN DRY RUN MODE.");
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                IEnumerable<SoloScore> importedScores = await conn.QueryAsync<SoloScore>(
+                    "SELECT `id`, `has_replay` FROM scores WHERE id BETWEEN @lastId AND (@lastId + @batchSize - 1) AND has_replay = 1 AND legacy_score_id IS NULL ORDER BY id",
+                    new
+                    {
+                        lastId,
+                        batchSize = BatchSize
+                    });
+
+                if (!importedScores.Any())
+                {
+                    if (lastId > await conn.QuerySingleAsync<ulong>("SELECT MAX(id) FROM scores"))
+                    {
+                        Console.WriteLine("All done!");
+                        break;
+                    }
+
+                    lastId += (ulong)BatchSize;
+                    continue;
+                }
+
+                foreach (var score in importedScores)
+                {
+                    if (!score.has_replay) continue;
+
+                    try
+                    {
+                        // check the replay is available on s3.
+                        _ = await s3.GetObjectMetadataAsync("score-replays", score.id.ToString(CultureInfo.InvariantCulture), cancellationToken);
```

Comment:
As already touched on above, this feels quite dodge. Any old network failure and this deletes replays. No error logging either. Is there really no better way to at least check for a slightly more specific error?

There's [some old code online that does this](https://github.com/aws/aws-sdk-net/blob/c688adc3850fbcbf708154258963afbb680b59d1/sdk/src/Services/S3/Custom/_bcl/IO/S3FileInfo.cs#L104-L147), but then again, not sure how up-to-date that is, given `S3FileInfo.Exists()` was apparently a method offered out of the box in AWS SDK, but now it doesn't seem to be there at all anymore...

Reply:
This was also just copypaste from our S3 helper utility in `osu-server-spectator`. But I agree it's pretty dodge.

Reply:
I'll give what you linked a try, but the exception being thown is `PermissionDenied`, which is the same for a multitude of cases including `NotFound`. I'm not sure we get any more fidelity in exceptions these days from S3.

Reply:
Sure, I'd still say matching on that is better than indiscriminate catching.